File: naion/csm.go
package naion

import (
	"crypto/ed25519"
	"crypto/rand"
	"crypto/sha512"
	"errors"
	"math/big"

	"golang.org/x/crypto/chacha20"
	"golang.org/x/crypto/chacha20poly1305"
	"golang.org/x/crypto/curve25519"
)

const (
	SignBytes   = ed25519.SignatureSize
	EdSeedBytes = ed25519.SeedSize
	EdPKBytes   = ed25519.PublicKeySize
	EdSKBytes   = ed25519.PrivateKeySize
	XPKBytes    = 32
	XSKBytes    = 32
	NonceBytes  = chacha20poly1305.NonceSizeX
	MACBytes    = 16

	MaxUDPDatagramBytes = 1024
)

const (
	packetFixedOverheadBytes = SignBytes + XPKBytes + NonceBytes + MACBytes
	clientPacketFixedBytes   = packetFixedOverheadBytes + EdPKBytes
	serverPacketFixedBytes   = packetFixedOverheadBytes
	MaxClientPayloadBytes    = MaxUDPDatagramBytes - clientPacketFixedBytes
	MaxServerPayloadBytes    = MaxUDPDatagramBytes - serverPacketFixedBytes
)

var (
	ErrInvalidArgument = errors.New("csm: invalid argument")
	ErrBufferTooSmall  = errors.New("csm: buffer too small")
	ErrCrypto          = errors.New("csm: crypto failed")
	ErrVerifyFailed    = errors.New("csm: verify failed")
	ErrState           = errors.New("csm: invalid state")
	ErrNoData          = errors.New("csm: empty data")
	ErrPayloadTooLarge = errors.New("csm: payload too large")
	ErrInvalidPacket   = errors.New("csm: invalid packet")
)

func Init() int { return 0 }

func ClientEncryptSize(plaintextLen int) int {
	if plaintextLen < 0 {
		return 0
	}
	return clientPacketFixedBytes + plaintextLen
}

func ClientDecryptMaxPlaintextSize(packetLen int) int {
	if packetLen <= serverPacketFixedBytes {
		return 0
	}
	return packetLen - serverPacketFixedBytes
}

func ServerEncryptSize(plaintextLen int) int {
	if plaintextLen < 0 {
		return 0
	}
	return serverPacketFixedBytes + plaintextLen
}

func ServerDecryptMaxPlaintextSize(packetLen int) int {
	if packetLen <= clientPacketFixedBytes {
		return 0
	}
	return packetLen - clientPacketFixedBytes
}

type Client struct {
	edSeed            [EdSeedBytes]byte
	edSecretKey       ed25519.PrivateKey
	edPublicKey       ed25519.PublicKey
	serverEdPublicKey ed25519.PublicKey
}

type Server struct {
	edSeed                [EdSeedBytes]byte
	edSecretKey           ed25519.PrivateKey
	edPublicKey           ed25519.PublicKey
	clientEdPublicKey     ed25519.PublicKey
	clientPublicKeyInited bool
}

func NewClient(edSeedClient []byte, edPublicKeyServer []byte) (*Client, error) {
	if len(edSeedClient) != EdSeedBytes || len(edPublicKeyServer) != EdPKBytes {
		return nil, ErrInvalidArgument
	}
	var seed [EdSeedBytes]byte
	copy(seed[:], edSeedClient)
	edSecretKey := ed25519.NewKeyFromSeed(seed[:])
	edPublicKey := edSecretKey.Public().(ed25519.PublicKey)

	serverPK := make([]byte, EdPKBytes)
	copy(serverPK, edPublicKeyServer)

	return &Client{
		edSeed:            seed,
		edSecretKey:       edSecretKey,
		edPublicKey:       append(ed25519.PublicKey(nil), edPublicKey...),
		serverEdPublicKey: append(ed25519.PublicKey(nil), serverPK...),
	}, nil
}

func NewServer(edSeedServer []byte) (*Server, error) {
	if len(edSeedServer) != EdSeedBytes {
		return nil, ErrInvalidArgument
	}
	var seed [EdSeedBytes]byte
	copy(seed[:], edSeedServer)
	edSecretKey := ed25519.NewKeyFromSeed(seed[:])
	edPublicKey := edSecretKey.Public().(ed25519.PublicKey)
	return &Server{
		edSeed:      seed,
		edSecretKey: edSecretKey,
		edPublicKey: append(ed25519.PublicKey(nil), edPublicKey...),
	}, nil
}

func (c *Client) Encrypt(plaintext []byte) ([]byte, error) {
	if c == nil {
		return nil, ErrInvalidArgument
	}
	if len(plaintext) == 0 {
		return nil, ErrNoData
	}
	if len(plaintext) > MaxClientPayloadBytes {
		return nil, ErrPayloadTooLarge
	}
	serverXPK, err := edPublicToCurve25519(c.serverEdPublicKey)
	if err != nil {
		return nil, err
	}
	sessionXPK, sessionXSK, err := generateX25519Keypair()
	if err != nil {
		return nil, err
	}

	payload := make([]byte, 0, EdPKBytes+len(plaintext))
	payload = append(payload, c.edPublicKey...)
	payload = append(payload, plaintext...)

	sealed, err := seal(payload, serverXPK, sessionXSK, sessionXPK[:])
	if err != nil {
		return nil, err
	}

	body := make([]byte, 0, XPKBytes+len(sealed))
	body = append(body, sessionXPK[:]...)
	body = append(body, sealed...)
	sig := ed25519.Sign(c.edSecretKey, body)
	out := make([]byte, 0, len(sig)+len(body))
	out = append(out, sig...)
	out = append(out, body...)
	return out, nil
}

func (c *Client) Decrypt(packet []byte) ([]byte, error) {
	if c == nil {
		return nil, ErrInvalidArgument
	}
	minSize := serverPacketFixedBytes
	if len(packet) <= minSize {
		return nil, ErrInvalidPacket
	}
	signature := packet[:SignBytes]
	body := packet[SignBytes:]
	if !ed25519.Verify(c.serverEdPublicKey, body, signature) {
		return nil, ErrVerifyFailed
	}
	var sessionXPK [XPKBytes]byte
	copy(sessionXPK[:], body[:XPKBytes])
	nonceCipher := body[XPKBytes:]
	clientXSK := edSecretToCurve25519(c.edSecretKey)
	opened, err := openBox(nonceCipher, &sessionXPK, &clientXSK, sessionXPK[:])
	if err != nil {
		return nil, err
	}
	if len(opened) == 0 {
		return nil, ErrInvalidPacket
	}
	return append([]byte(nil), opened...), nil
}

func (s *Server) Decrypt(packet []byte) ([]byte, error) {
	if s == nil {
		return nil, ErrInvalidArgument
	}
	minSize := clientPacketFixedBytes
	if len(packet) <= minSize {
		return nil, ErrInvalidPacket
	}
	signature := packet[:SignBytes]
	body := packet[SignBytes:]
	var sessionXPK [XPKBytes]byte
	copy(sessionXPK[:], body[:XPKBytes])
	nonceCipher := body[XPKBytes:]
	serverXSK := edSecretToCurve25519(s.edSecretKey)
	opened, err := openBox(nonceCipher, &sessionXPK, &serverXSK, sessionXPK[:])
	if err != nil {
		return nil, err
	}

	if len(opened) <= EdPKBytes {
		return nil, ErrInvalidPacket
	}

	clientEDPK := append([]byte(nil), opened[:EdPKBytes]...)
	plaintext := append([]byte(nil), opened[EdPKBytes:]...)
	if !ed25519.Verify(clientEDPK, body, signature) {
		return nil, ErrVerifyFailed
	}
	s.clientEdPublicKey = clientEDPK
	s.clientPublicKeyInited = true
	return plaintext, nil
}

func (s *Server) Encrypt(plaintext []byte) ([]byte, error) {
	if s == nil {
		return nil, ErrInvalidArgument
	}
	if !s.clientPublicKeyInited {
		return nil, ErrState
	}
	if len(plaintext) == 0 {
		return nil, ErrNoData
	}
	if len(plaintext) > MaxServerPayloadBytes {
		return nil, ErrPayloadTooLarge
	}
	clientXPK, err := edPublicToCurve25519(s.clientEdPublicKey)
	if err != nil {
		return nil, err
	}
	sessionXPK, sessionXSK, err := generateX25519Keypair()
	if err != nil {
		return nil, err
	}

	sealed, err := seal(plaintext, clientXPK, sessionXSK, sessionXPK[:])
	if err != nil {
		return nil, err
	}
	body := make([]byte, 0, XPKBytes+len(sealed))
	body = append(body, sessionXPK[:]...)
	body = append(body, sealed...)
	sig := ed25519.Sign(s.edSecretKey, body)
	out := make([]byte, 0, len(sig)+len(body))
	out = append(out, sig...)
	out = append(out, body...)
	return out, nil
}

func (s *Server) ClientPublicKeyInitialized() bool {
	return s != nil && s.clientPublicKeyInited
}

func (s *Server) ClientEdPublicKey() []byte {
	if s == nil || !s.clientPublicKeyInited {
		return nil
	}
	return append([]byte(nil), s.clientEdPublicKey...)
}

func (c *Client) EdPublicKey() []byte {
	if c == nil {
		return nil
	}
	return append([]byte(nil), c.edPublicKey...)
}

func (s *Server) EdPublicKey() []byte {
	if s == nil {
		return nil
	}
	return append([]byte(nil), s.edPublicKey...)
}

func generateX25519Keypair() (*[XPKBytes]byte, *[XSKBytes]byte, error) {
	var sk [XSKBytes]byte
	if _, err := rand.Read(sk[:]); err != nil {
		return nil, nil, ErrCrypto
	}
	clampCurve25519Secret(&sk)
	pkSlice, err := curve25519.X25519(sk[:], curve25519.Basepoint)
	if err != nil {
		return nil, nil, ErrCrypto
	}
	var pk [XPKBytes]byte
	copy(pk[:], pkSlice)
	return &pk, &sk, nil
}

func seal(plaintext []byte, peerXPK *[XPKBytes]byte, selfXSK *[XSKBytes]byte, aad []byte) ([]byte, error) {
	if len(plaintext) == 0 || peerXPK == nil || selfXSK == nil {
		return nil, ErrInvalidArgument
	}
	var nonce [NonceBytes]byte
	if _, err := rand.Read(nonce[:]); err != nil {
		return nil, ErrCrypto
	}
	aeadKey, err := deriveAEADKey(peerXPK, selfXSK)
	if err != nil {
		return nil, ErrCrypto
	}
	aead, err := chacha20poly1305.NewX(aeadKey)
	if err != nil {
		return nil, ErrCrypto
	}
	ciphertextAndTag := aead.Seal(nil, nonce[:], plaintext, aad)
	out := make([]byte, 0, NonceBytes+len(ciphertextAndTag))
	out = append(out, nonce[:]...)
	out = append(out, ciphertextAndTag[len(ciphertextAndTag)-MACBytes:]...)
	out = append(out, ciphertextAndTag[:len(ciphertextAndTag)-MACBytes]...)
	return out, nil
}

func openBox(nonceCipher []byte, peerXPK *[XPKBytes]byte, selfXSK *[XSKBytes]byte, aad []byte) ([]byte, error) {
	if len(nonceCipher) <= NonceBytes+MACBytes || peerXPK == nil || selfXSK == nil {
		return nil, ErrInvalidArgument
	}
	var nonce [NonceBytes]byte
	copy(nonce[:], nonceCipher[:NonceBytes])
	macCiphertext := nonceCipher[NonceBytes:]
	aeadKey, err := deriveAEADKey(peerXPK, selfXSK)
	if err != nil {
		return nil, ErrCrypto
	}
	aead, err := chacha20poly1305.NewX(aeadKey)
	if err != nil {
		return nil, ErrCrypto
	}
	combined := make([]byte, 0, len(macCiphertext))
	combined = append(combined, macCiphertext[MACBytes:]...)
	combined = append(combined, macCiphertext[:MACBytes]...)
	opened, err := aead.Open(nil, nonce[:], combined, aad)
	if err != nil {
		return nil, ErrCrypto
	}
	return opened, nil
}

func deriveAEADKey(peerXPK *[XPKBytes]byte, selfXSK *[XSKBytes]byte) ([]byte, error) {
	shared, err := curve25519.X25519(selfXSK[:], peerXPK[:])
	if err != nil {
		return nil, ErrCrypto
	}
	zeroNonce := make([]byte, 16)
	key, err := chacha20.HChaCha20(shared, zeroNonce)
	if err != nil {
		return nil, ErrCrypto
	}
	return key, nil
}

func edSecretToCurve25519(edSecret ed25519.PrivateKey) [XSKBytes]byte {
	seed := edSecret.Seed()
	h := sha512.Sum512(seed)
	h[0] &= 248
	h[31] &= 127
	h[31] |= 64
	var out [XSKBytes]byte
	copy(out[:], h[:32])
	return out
}

func edPublicToCurve25519(edPublic ed25519.PublicKey) (*[XPKBytes]byte, error) {
	if len(edPublic) != EdPKBytes {
		return nil, ErrInvalidArgument
	}
	const pHex = "7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffed"
	p, ok := new(big.Int).SetString(pHex, 16)
	if !ok {
		return nil, ErrCrypto
	}

	yBytes := make([]byte, 32)
	copy(yBytes, edPublic)
	yBytes[31] &= 0x7f
	reverse(yBytes)
	y := new(big.Int).SetBytes(yBytes)
	if y.Cmp(p) >= 0 {
		return nil, ErrCrypto
	}

	one := big.NewInt(1)
	num := new(big.Int).Add(one, y)
	num.Mod(num, p)
	den := new(big.Int).Sub(one, y)
	den.Mod(den, p)
	inv := new(big.Int).ModInverse(den, p)
	if inv == nil {
		return nil, ErrCrypto
	}
	u := new(big.Int).Mul(num, inv)
	u.Mod(u, p)

	be := u.Bytes()
	le := make([]byte, 32)
	copy(le[32-len(be):], be)
	reverse(le)
	var out [XPKBytes]byte
	copy(out[:], le)
	return &out, nil
}

func clampCurve25519Secret(sk *[XSKBytes]byte) {
	sk[0] &= 248
	sk[31] &= 127
	sk[31] |= 64
}

func reverse(b []byte) {
	for i, j := 0, len(b)-1; i < j; i, j = i+1, j-1 {
		b[i], b[j] = b[j], b[i]
	}
}

File: naion/csm.py
"""
CSM = client-server-model
Updated runtime secure messaging helper aligned with the C++ and Go protocol.

Model:
- Client holds server Ed25519 public key.
- Server does not know client key material until the first client packet is decrypted.

Packet format:
- outer packet (both directions):
  signature(64) | session_x25519_pk(32) | nonce(24) | mac(16) | ciphertext

- client -> server plaintext before seal:
  client_ed25519_pk(32) | app_plaintext

- server -> client plaintext before seal:
  app_plaintext

Cipher suite:
- Ed25519 detached signature
- X25519 shared key agreement
- XChaCha20-Poly1305-IETF with session_x25519_pk as AAD
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional, Tuple

from nacl import bindings
from nacl.exceptions import BadSignatureError, CryptoError
from nacl.utils import random as random_bytes


SIGN_BYTES = bindings.crypto_sign_BYTES
ED_SEED_BYTES = bindings.crypto_sign_SEEDBYTES
ED_PK_BYTES = bindings.crypto_sign_PUBLICKEYBYTES
ED_SK_BYTES = bindings.crypto_sign_SECRETKEYBYTES
X_PK_BYTES = bindings.crypto_box_PUBLICKEYBYTES
X_SK_BYTES = bindings.crypto_box_SECRETKEYBYTES
NONCE_BYTES = bindings.crypto_box_NONCEBYTES
MAC_BYTES = getattr(bindings, "crypto_box_MACBYTES", 16)

MAX_UDP_DATAGRAM_BYTES = 1024
PACKET_FIXED_OVERHEAD_BYTES = SIGN_BYTES + X_PK_BYTES + NONCE_BYTES + MAC_BYTES
CLIENT_PACKET_FIXED_BYTES = PACKET_FIXED_OVERHEAD_BYTES + ED_PK_BYTES
SERVER_PACKET_FIXED_BYTES = PACKET_FIXED_OVERHEAD_BYTES
MAX_CLIENT_PAYLOAD_BYTES = MAX_UDP_DATAGRAM_BYTES - CLIENT_PACKET_FIXED_BYTES
MAX_SERVER_PAYLOAD_BYTES = MAX_UDP_DATAGRAM_BYTES - SERVER_PACKET_FIXED_BYTES
DEFAULT_REPLAY_RETENTION_MS = 5 * 60 * 1000


class CSMError(Exception):
    pass


class VerifyFailed(CSMError):
    pass


class InvalidPacket(CSMError):
    pass


class StateError(CSMError):
    pass


class PayloadTooLarge(CSMError):
    pass


def init() -> int:
    return bindings.sodium_init()


def client_encrypt_size(plaintext_len: int) -> int:
    return 0 if plaintext_len < 0 else CLIENT_PACKET_FIXED_BYTES + plaintext_len


def client_decrypt_max_plaintext_size(packet_len: int) -> int:
    return 0 if packet_len <= SERVER_PACKET_FIXED_BYTES else packet_len - SERVER_PACKET_FIXED_BYTES


def server_encrypt_size(plaintext_len: int) -> int:
    return 0 if plaintext_len < 0 else SERVER_PACKET_FIXED_BYTES + plaintext_len


def server_decrypt_max_plaintext_size(packet_len: int) -> int:
    return 0 if packet_len <= CLIENT_PACKET_FIXED_BYTES else packet_len - CLIENT_PACKET_FIXED_BYTES


def _ensure_size(name: str, value: bytes, expected: int) -> None:
    if len(value) != expected:
        raise CSMError(f"{name} size mismatch: got {len(value)}, expected {expected}")


def _sign_detached(message: bytes, ed_secret_key: bytes) -> bytes:
    _ensure_size("ed_secret_key", ed_secret_key, ED_SK_BYTES)
    signed = bindings.crypto_sign(message, ed_secret_key)
    return signed[:SIGN_BYTES]


def _verify_detached(signature: bytes, message: bytes, ed_public_key: bytes) -> bool:
    _ensure_size("signature", signature, SIGN_BYTES)
    _ensure_size("ed_public_key", ed_public_key, ED_PK_BYTES)
    try:
        opened = bindings.crypto_sign_open(signature + message, ed_public_key)
        return opened == message
    except BadSignatureError:
        return False


def _ed_pk_to_xpk(ed_public_key: bytes) -> bytes:
    _ensure_size("ed_public_key", ed_public_key, ED_PK_BYTES)
    return bindings.crypto_sign_ed25519_pk_to_curve25519(ed_public_key)


def _ed_sk_to_xsk(ed_secret_key: bytes) -> bytes:
    _ensure_size("ed_secret_key", ed_secret_key, ED_SK_BYTES)
    return bindings.crypto_sign_ed25519_sk_to_curve25519(ed_secret_key)


def _x25519_keypair() -> Tuple[bytes, bytes]:
    return bindings.crypto_box_keypair()


def _rotl32(value: int, bits: int) -> int:
    return ((value << bits) & 0xFFFFFFFF) | (value >> (32 - bits))


def _quarterround(state: list[int], a: int, b: int, c: int, d: int) -> None:
    state[a] = (state[a] + state[b]) & 0xFFFFFFFF
    state[d] = _rotl32(state[d] ^ state[a], 16)
    state[c] = (state[c] + state[d]) & 0xFFFFFFFF
    state[b] = _rotl32(state[b] ^ state[c], 12)
    state[a] = (state[a] + state[b]) & 0xFFFFFFFF
    state[d] = _rotl32(state[d] ^ state[a], 8)
    state[c] = (state[c] + state[d]) & 0xFFFFFFFF
    state[b] = _rotl32(state[b] ^ state[c], 7)


def _hchacha20(key: bytes, nonce16: bytes) -> bytes:
    _ensure_size("hchacha20_key", key, 32)
    _ensure_size("hchacha20_nonce", nonce16, 16)
    constants = b"expand 32-byte k"
    words = [
        int.from_bytes(constants[i : i + 4], "little")
        for i in range(0, 16, 4)
    ]
    words.extend(int.from_bytes(key[i : i + 4], "little") for i in range(0, 32, 4))
    words.extend(int.from_bytes(nonce16[i : i + 4], "little") for i in range(0, 16, 4))
    for _ in range(10):
        _quarterround(words, 0, 4, 8, 12)
        _quarterround(words, 1, 5, 9, 13)
        _quarterround(words, 2, 6, 10, 14)
        _quarterround(words, 3, 7, 11, 15)
        _quarterround(words, 0, 5, 10, 15)
        _quarterround(words, 1, 6, 11, 12)
        _quarterround(words, 2, 7, 8, 13)
        _quarterround(words, 3, 4, 9, 14)
    out_words = words[:4] + words[12:16]
    return b"".join(word.to_bytes(4, "little") for word in out_words)


def _derive_aead_key(peer_xpk: bytes, self_xsk: bytes) -> bytes:
    shared = bindings.crypto_scalarmult(self_xsk, peer_xpk)
    return _hchacha20(shared, b"\x00" * 16)


def _seal(plaintext: bytes, peer_xpk: bytes, self_xsk: bytes, aad: bytes) -> bytes:
    if not plaintext:
        raise CSMError("seal plaintext is empty")
    nonce = random_bytes(NONCE_BYTES)
    _ensure_size("peer_xpk", peer_xpk, X_PK_BYTES)
    _ensure_size("self_xsk", self_xsk, X_SK_BYTES)
    ekey = _derive_aead_key(peer_xpk, self_xsk)
    encrypted = bindings.crypto_aead_xchacha20poly1305_ietf_encrypt(plaintext, aad, nonce, ekey)
    return nonce + encrypted[-MAC_BYTES:] + encrypted[:-MAC_BYTES]


def _open(nonce_cipher: bytes, peer_xpk: bytes, self_xsk: bytes, aad: bytes) -> bytes:
    if len(nonce_cipher) <= NONCE_BYTES + MAC_BYTES:
        raise InvalidPacket("ciphertext too short")
    nonce = nonce_cipher[:NONCE_BYTES]
    mac = nonce_cipher[NONCE_BYTES : NONCE_BYTES + MAC_BYTES]
    ciphertext = nonce_cipher[NONCE_BYTES + MAC_BYTES :]
    _ensure_size("peer_xpk", peer_xpk, X_PK_BYTES)
    _ensure_size("self_xsk", self_xsk, X_SK_BYTES)
    try:
        ekey = _derive_aead_key(peer_xpk, self_xsk)
        return bindings.crypto_aead_xchacha20poly1305_ietf_decrypt(ciphertext + mac, aad, nonce, ekey)
    except CryptoError as e:
        raise CSMError("open failed") from e


@dataclass
class Client:
    ed_seed: bytes
    ed_secret_key: bytes
    ed_public_key: bytes
    server_ed_public_key: bytes

    @classmethod
    def create(cls, ed_seed_client: bytes, ed_public_key_server: bytes) -> "Client":
        _ensure_size("ed_seed_client", ed_seed_client, ED_SEED_BYTES)
        _ensure_size("ed_public_key_server", ed_public_key_server, ED_PK_BYTES)
        ed_public_key, ed_secret_key = bindings.crypto_sign_seed_keypair(ed_seed_client)
        return cls(ed_seed_client, ed_secret_key, ed_public_key, ed_public_key_server)

    def encrypt(self, plaintext: bytes) -> bytes:
        if not plaintext:
            raise CSMError("client encrypt empty data")
        if len(plaintext) > MAX_CLIENT_PAYLOAD_BYTES:
            raise PayloadTooLarge("client payload exceeds UDP budget")
        server_xpk = _ed_pk_to_xpk(self.server_ed_public_key)
        session_xpk, session_xsk = _x25519_keypair()
        payload = self.ed_public_key + plaintext
        nonce_cipher = _seal(payload, server_xpk, session_xsk, session_xpk)
        body = session_xpk + nonce_cipher
        sig = _sign_detached(body, self.ed_secret_key)
        return sig + body

    def decrypt(self, packet: bytes) -> bytes:
        min_size = SERVER_PACKET_FIXED_BYTES
        if len(packet) <= min_size:
            raise InvalidPacket("client decrypt packet too short")
        signature = packet[:SIGN_BYTES]
        body = packet[SIGN_BYTES:]
        if not _verify_detached(signature, body, self.server_ed_public_key):
            raise VerifyFailed("client decrypt signature verify failed")
        session_xpk = body[:X_PK_BYTES]
        nonce_cipher = body[X_PK_BYTES:]
        client_xsk = _ed_sk_to_xsk(self.ed_secret_key)
        plaintext = _open(nonce_cipher, session_xpk, client_xsk, session_xpk)
        if not plaintext:
            raise InvalidPacket("client decrypt payload missing")
        return plaintext


@dataclass
class Server:
    ed_seed: bytes
    ed_secret_key: bytes
    ed_public_key: bytes
    client_ed_public_key: Optional[bytes] = None

    @classmethod
    def create(cls, ed_seed_server: bytes) -> "Server":
        _ensure_size("ed_seed_server", ed_seed_server, ED_SEED_BYTES)
        ed_public_key, ed_secret_key = bindings.crypto_sign_seed_keypair(ed_seed_server)
        return cls(ed_seed_server, ed_secret_key, ed_public_key, None)

    @property
    def client_public_key_initialized(self) -> bool:
        return self.client_ed_public_key is not None

    def decrypt(self, packet: bytes) -> bytes:
        min_size = CLIENT_PACKET_FIXED_BYTES
        if len(packet) <= min_size:
            raise InvalidPacket("server decrypt packet too short")
        signature = packet[:SIGN_BYTES]
        body = packet[SIGN_BYTES:]
        session_xpk = body[:X_PK_BYTES]
        nonce_cipher = body[X_PK_BYTES:]
        server_xsk = _ed_sk_to_xsk(self.ed_secret_key)
        opened = _open(nonce_cipher, session_xpk, server_xsk, session_xpk)
        if len(opened) <= ED_PK_BYTES:
            raise InvalidPacket("server decrypt opened payload too short")
        client_ed_pk = opened[:ED_PK_BYTES]
        plaintext = opened[ED_PK_BYTES:]
        if not _verify_detached(signature, body, client_ed_pk):
            raise VerifyFailed("server decrypt signature verify failed")
        self.client_ed_public_key = client_ed_pk
        return plaintext

    def encrypt(self, plaintext: bytes) -> bytes:
        if not self.client_public_key_initialized:
            raise StateError("server encrypt client public key is not initialized")
        if not plaintext:
            raise CSMError("server encrypt empty data")
        if len(plaintext) > MAX_SERVER_PAYLOAD_BYTES:
            raise PayloadTooLarge("server payload exceeds UDP budget")
        assert self.client_ed_public_key is not None
        client_xpk = _ed_pk_to_xpk(self.client_ed_public_key)
        session_xpk, session_xsk = _x25519_keypair()
        nonce_cipher = _seal(plaintext, client_xpk, session_xsk, session_xpk)
        body = session_xpk + nonce_cipher
        sig = _sign_detached(body, self.ed_secret_key)
        return sig + body
